File: cpp/base/src/main/vsdk/toolkit/render/raster/Rasterizer2D.cpp
#include <cmath>

#include "java/lang/Math.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/common/VSDKFatalException.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/media/Image.h"
#include "vsdk/toolkit/media/RGBPixel.h"
#include "vsdk/toolkit/environment/geometry/element/Vertex2D.h"
#include "vsdk/toolkit/environment/geometry/surface/polygon/_Polygon2DContour.h"
#include "vsdk/toolkit/render/raster/Rasterizer2D.h"
static void quicksortDoubles(double* array, int left0, int right0)
{
    int left = left0;
    int right = right0 + 1;
    double pivot;
    double temp;

    pivot = array[left0];
    do {
        do {
            left++;
        } while (left <= right0 && array[left] < pivot);
        do {
            right--;
            if (right < 0) {
                Logger::reportMessage("Rasterizer2D", Logger::ERROR, "quicksortDoubles", "quicksort span underflow");
                throw VSDKFatalException("quicksort span underflow");
            }
        } while (array[right] > pivot);
        if (left < right) {
            temp = array[left];
            array[left] = array[right];
            array[right] = temp;
        }
    } while (left <= right);

    temp = array[left0];
    array[left0] = array[right];
    array[right] = temp;

    if (left0 < right) {
        quicksortDoubles(array, left0, right);
    }
    if (left < right0) {
        quicksortDoubles(array, left, right0);
    }
}

static void sortDoubles(java::ArrayList<double>& arr)
{
    if (arr.size() < 2) {
        return;
    }
    quicksortDoubles(arr.data(), 0, static_cast<int>(arr.size()) - 1);
}

void Rasterizer2D::sortFillEdges(java::ArrayList<FillEdge>& edges)
{
    for (size_t i = 1; i < edges.size(); ++i) {
        FillEdge key = edges[i];
        size_t j = i;
        while (j > 0) {
            const FillEdge& prev = edges[j - 1];
            bool shouldMove =
                prev.xAtCurrentY > key.xAtCurrentY ||
                (prev.xAtCurrentY == key.xAtCurrentY &&
                    (prev.inverseSlope > key.inverseSlope ||
                        (prev.inverseSlope == key.inverseSlope &&
                            prev.sortOrder > key.sortOrder)));
            if (!shouldMove) {
                break;
            }
            edges[j] = prev;
            --j;
        }
        edges[j] = key;
    }
}

void Rasterizer2D::drawLine(Image* img, int x0, int y0, int x1, int y1, const RGBPixel& p)
{
    if (img == 0) return;
    double dx = (double)(x1 - x0);
    double dy = (double)(y1 - y0);
    int x, y;
    double xx, yy;

    if (std::abs(dx) > VSDK::EPSILON && std::abs(dy / dx) <= 1.0 && x1 > x0) {
        double dydx = dy / dx;
        for (x = x0, yy = (double)y0; x <= x1; x++) {
            y = (int)yy;
            if (x >= 0 && x < img->getXSize() && y >= 0 && y < img->getYSize()) {
                RGBPixel c(p);
                img->putPixelRgb(x, y, &c);
            }
            yy += dydx;
        }
    }
    else if (std::abs(dx) > VSDK::EPSILON && std::abs(dy / dx) <= 1.0 && x1 < x0) {
        double dydx = dy / dx;
        for (x = x1, yy = (double)y1; x <= x0; x++) {
            y = (int)yy;
            if (x >= 0 && x < img->getXSize() && y >= 0 && y < img->getYSize()) {
                RGBPixel c(p);
                img->putPixelRgb(x, y, &c);
            }
            yy += dydx;
        }
    }
    else if (std::abs(dy) > VSDK::EPSILON && y1 > y0) {
        double dxdy = dx / dy;
        for (y = y0, xx = (double)x0; y <= y1; y++) {
            x = (int)xx;
            if (x >= 0 && x < img->getXSize() && y >= 0 && y < img->getYSize()) {
                RGBPixel c(p);
                img->putPixelRgb(x, y, &c);
            }
            xx += dxdy;
        }
    }
    else if (std::abs(dy) > VSDK::EPSILON && y1 < y0) {
        double dxdy = dx / dy;
        for (y = y1, xx = (double)x1; y <= y0; y++) {
            x = (int)xx;
            if (x >= 0 && x < img->getXSize() && y >= 0 && y < img->getYSize()) {
                RGBPixel c(p);
                img->putPixelRgb(x, y, &c);
            }
            xx += dxdy;
        }
    }
}

void Rasterizer2D::drawPolygon(Image* img, Polygon2D& p, const RGBPixel& color)
{
    if (img == 0) return;
    for (long int i = 0; i < p.loops.size(); i++) {
        _Polygon2DContour* contour = p.loops[i];
        if (contour->vertices.size() < 2) continue;
        Vertex2D va(0, 0), vb(0, 0);
        for (long int j = 0; j + 1 < contour->vertices.size(); j++) {
            va = contour->vertices[j];
            vb = contour->vertices[j + 1];
            drawLine(img, (int)va.x, (int)va.y, (int)vb.x, (int)vb.y, color);
        }
        va = vb;
        vb = contour->vertices[0];
        drawLine(img, (int)va.x, (int)va.y, (int)vb.x, (int)vb.y, color);
    }
}

int Rasterizer2D::clamp(int value, int minValue, int maxValue)
{
    if (value < minValue) {
        return minValue;
    }
    if (value > maxValue) {
        return maxValue;
    }
    return value;
}

void Rasterizer2D::addFillEdge(java::ArrayList< java::ArrayList<FillEdge> >& buckets,
    const Vertex2D& a, const Vertex2D& b, int imageHeight, int yRange[2],
    int sortOrder)
{
    double dx = b.x - a.x;
    double dy = b.y - a.y;

    if (std::abs(dx) < VSDK::EPSILON && std::abs(dy) < VSDK::EPSILON) {
        return;
    }
    if (std::abs(dy) < VSDK::EPSILON) {
        return;
    }

    const Vertex2D* top = &a;
    const Vertex2D* bottom = &b;
    if (a.y > b.y) {
        top = &b;
        bottom = &a;
        dx = -dx;
        dy = -dy;
    }

    double inverseSlope = dx / dy;
    int yMin = (int)std::ceil(top->y);
    int yMaxExclusive = (int)std::ceil(bottom->y);

    if (yMin >= yMaxExclusive) {
        return;
    }

    int clippedYMin = clamp(yMin, 0, imageHeight);
    int clippedYMaxExclusive = clamp(yMaxExclusive, 0, imageHeight);

    if (clippedYMin >= clippedYMaxExclusive) {
        return;
    }

    FillEdge edge;
    edge.yMin = clippedYMin;
    edge.yMaxExclusive = clippedYMaxExclusive;
    edge.inverseSlope = inverseSlope;
    edge.xAtCurrentY = top->x + (((double)clippedYMin) - top->y) * inverseSlope;
    edge.sortOrder = sortOrder;

    buckets[clippedYMin].add(edge);
    if (clippedYMin < yRange[0]) {
        yRange[0] = clippedYMin;
    }
    if (clippedYMaxExclusive > yRange[1]) {
        yRange[1] = clippedYMaxExclusive;
    }
}

void Rasterizer2D::rasterizePolygonSpans(Image* img, Polygon2D& polygon,
    SpanShader& shader)
{
    int imageWidth = img->getXSize();
    int imageHeight = img->getYSize();

    if (imageWidth <= 0 || imageHeight <= 0) {
        return;
    }

    java::ArrayList< java::ArrayList<FillEdge> > buckets(imageHeight);
    for (int i = 0; i < imageHeight; i++) {
        buckets.add(java::ArrayList<FillEdge>());
    }
    int yRange[2] = {imageHeight, 0};
    int sortOrder = 0;

    for (long int i = 0; i < polygon.loops.size(); i++) {
        _Polygon2DContour* contour = polygon.loops[i];
        long int vertexCount = contour->vertices.size();
        if (vertexCount < 2) {
            continue;
        }

        for (long int j = 0; j < vertexCount; j++) {
            const Vertex2D& a = contour->vertices[j];
            const Vertex2D& b = contour->vertices[(j + 1) % vertexCount];
            addFillEdge(buckets, a, b, imageHeight, yRange, sortOrder);
            sortOrder++;
        }
    }

    if (yRange[0] >= yRange[1]) {
        return;
    }

    java::ArrayList<FillEdge> activeEdges;

    for (int y = yRange[0]; y < yRange[1]; y++) {
        java::ArrayList<FillEdge>& bucket = buckets[y];
        for (long int i = 0; i < bucket.size(); i++) {
            activeEdges.add(bucket[i]);
        }

        for (int i = (int)activeEdges.size() - 1; i >= 0; i--) {
            if (y >= activeEdges[(size_t)i].yMaxExclusive) {
                activeEdges.remove(i);
            }
        }

        if (activeEdges.size() < 2) {
            for (size_t i = 0; i < activeEdges.size(); i++) {
                activeEdges[i].xAtCurrentY += activeEdges[i].inverseSlope;
            }
            continue;
        }

        sortFillEdges(activeEdges);

        for (size_t i = 0; i + 1 < activeEdges.size(); i += 2) {
            double xLeft = activeEdges[i].xAtCurrentY;
            double xRight = activeEdges[i + 1].xAtCurrentY;

            if (xLeft > xRight) {
                double tmp = xLeft;
                xLeft = xRight;
                xRight = tmp;
            }

            int xStart = (int)std::ceil(xLeft);
            int xEndExclusive = (int)std::ceil(xRight);

            if (xEndExclusive <= 0 || xStart >= imageWidth) {
                continue;
            }

            xStart = clamp(xStart, 0, imageWidth);
            xEndExclusive = clamp(xEndExclusive, 0, imageWidth);

            if (xStart < xEndExclusive) {
                shader.shade(img, polygon, y, xStart, xEndExclusive);
            }
        }

        for (size_t i = 0; i < activeEdges.size(); i++) {
            activeEdges[i].xAtCurrentY += activeEdges[i].inverseSlope;
        }
    }
}

void Rasterizer2D::fillPolygon(Image* img, Polygon2D& p, const RGBPixel& color)
{
    if (img == 0) return;
    class SolidSpanShader : public SpanShader {
    public:
        explicit SolidSpanShader(const RGBPixel& color) : color(color) {}

        virtual void shade(Image* img, Polygon2D& polygon, int y, int xStart,
            int xEndExclusive)
        {
            (void)polygon;
            for (int x = xStart; x < xEndExclusive; x++) {
                RGBPixel c(color);
                img->putPixelRgb(x, y, &c);
            }
        }

    private:
        RGBPixel color;
    };

    SolidSpanShader shader(color);
    rasterizePolygonSpans(img, p, shader);
}

void Rasterizer2D::fillSmoothPolygonCalculateColor(
    Polygon2D& p, double x, double y, RGBPixel& outPixel)
{
    double outR = 0.0, outG = 0.0, outB = 0.0;
    double totaldistance = 0.0;

    for (long int i = 0; i < p.loops.size(); i++) {
        for (long int j = 0; j < p.loops[i]->vertices.size(); j++) {
            const Vertex2D& va = p.loops[i]->vertices[j];
            double distance = 1.0 / (1.0 + std::sqrt((va.x - x) * (va.x - x) + (va.y - y) * (va.y - y)));
            totaldistance += distance;
            outR += va.color.r() * distance;
            outG += va.color.g() * distance;
            outB += va.color.b() * distance;
        }
    }
    if (totaldistance <= VSDK::EPSILON) {
        outPixel.r = 0; outPixel.g = 0; outPixel.b = 0;
        return;
    }
    double normalizedR = outR / totaldistance;
    double normalizedG = outG / totaldistance;
    double normalizedB = outB / totaldistance;
    double clippedR = java::Math::max(0.0, java::Math::min(1.0, normalizedR));
    double clippedG = java::Math::max(0.0, java::Math::min(1.0, normalizedG));
    double clippedB = java::Math::max(0.0, java::Math::min(1.0, normalizedB));
    int rr = (int)(clippedR * 255.0);
    int gg = (int)(clippedG * 255.0);
    int bb = (int)(clippedB * 255.0);
    outPixel.r = (char)(rr & 0xFF);
    outPixel.g = (char)(gg & 0xFF);
    outPixel.b = (char)(bb & 0xFF);
}

void Rasterizer2D::fillSmoothPolygon(Image* img, Polygon2D& p)
{
    if (img == 0) return;
    class SmoothSpanShader : public SpanShader {
    public:
        virtual void shade(Image* img, Polygon2D& polygon, int y, int xStart,
            int xEndExclusive)
        {
            RGBPixel color;
            for (int x = xStart; x < xEndExclusive; x++) {
                Rasterizer2D::fillSmoothPolygonCalculateColor(
                    polygon, x, y, color);
                img->putPixelRgb(x, y, &color);
            }
        }
    };

    SmoothSpanShader shader;
    rasterizePolygonSpans(img, p, shader);
}

File: typescript/base/src/main/vsdk/toolkit/render/raster/Rasterizer2D.ts
//= References:                                                             =
//= [BRES1965] Bresenham, J.E. "Algorithm for computer control of a digital =
//=            plotter" IBM Syst. J. 4, 1 (1965), 25-30.                    =
//= [FOLE1992] Foley, vanDam, Feiner, Hughes. "Computer Graphics, princi-   =
//=            ples and practice" - second edition, Addison Wesley, 1992.   =

import { Collections } from "../../../../java/util/Collections.js";
import type { Comparator } from "../../../../java/util/Comparator.js";
import { Double } from "../../../../java/lang/Double.js";
import { Integer } from "../../../../java/lang/Integer.js";
import { VSDK } from "../../common/VSDK.js";
import { Vertex2D } from "../../environment/geometry/element/Vertex2D.js";
import { RGBPixel } from "../../media/RGBPixel.js";
import { Image } from "../../media/Image.js";
import { Polygon2D } from "../../environment/geometry/surface/polygon/Polygon2D.js";
import { _Polygon2DContour } from "../../environment/geometry/surface/polygon/_Polygon2DContour.js";
import { RenderingElement } from "../RenderingElement.js";

class FillEdge {
    public yMin = 0;
    public yMaxExclusive = 0;
    public xAtCurrentY = 0;
    public inverseSlope = 0;
    public sortOrder = 0;
}

interface SpanShader {
    shade(img: Image, polygon: Polygon2D, y: number, xStart: number, xEndExclusive: number): void;
}

export class Rasterizer2D extends RenderingElement {
    private static readonly ACTIVE_EDGE_COMPARATOR: Comparator<FillEdge> = {
        compare(a: FillEdge, b: FillEdge): number {
            let cmp = Double.compare(a.xAtCurrentY, b.xAtCurrentY);
            if (cmp !== 0) {
                return cmp;
            }

            cmp = Double.compare(a.inverseSlope, b.inverseSlope);
            if (cmp !== 0) {
                return cmp;
            }

            return Integer.compare(a.sortOrder, b.sortOrder);
        },
    };

    /**
    This algorithm implements the Bresenham line algoritm with NO CLIPPING!
    See [BRES1965].
    Note that this is a currently naive implementation that makes use
    of double floating point arithmetic, while original Bresenham algorithm
    make use of most efficient integer line arithmetic.
    */
    public static drawLine(img: Image, x0: number, y0: number, x1: number, y1: number, p: RGBPixel): void {
        let dx: number, dy: number;
        let dxdy: number;
        let dydx: number;
        let x: number, y: number;
        let xx: number, yy: number;

        dx = x1 - x0;
        dy = y1 - y0;

        if (Math.abs(dx) > VSDK.EPSILON && Math.abs(dy / dx) <= 1 && x1 > x0) {
            // Pendiente entre -1 y 1
            dydx = dy / dx;
            for (x = x0, yy = y0; x <= x1; x++) {
                y = Math.trunc(yy);
                if (x >= 0 && x < img.getXSize() && y >= 0 && y < img.getYSize()) {
                    img.putPixelRgb(x, y, p);
                }
                yy += dydx;
            }
        } else if (Math.abs(dx) > VSDK.EPSILON && Math.abs(dy / dx) <= 1 && x1 < x0) {
            // Pendiente entre -1 y 1
            dydx = dy / dx;
            for (x = x1, yy = y1; x <= x0; x++) {
                y = Math.trunc(yy);
                if (x >= 0 && x < img.getXSize() && y >= 0 && y < img.getYSize()) {
                    img.putPixelRgb(x, y, p);
                }
                yy += dydx;
            }
        } else if (Math.abs(dy) > VSDK.EPSILON && y1 > y0) {
            // Pendiente mayor a 1 o menor a -1
            dxdy = dx / dy;
            for (y = y0, xx = x0; y <= y1; y++) {
                x = Math.trunc(xx);
                if (x >= 0 && x < img.getXSize() && y >= 0 && y < img.getYSize()) {
                    img.putPixelRgb(x, y, p);
                }
                xx += dxdy;
            }
        } else if (Math.abs(dy) > VSDK.EPSILON && y1 < y0) {
            // Pendiente mayor a 1 o menor a -1
            dxdy = dx / dy;
            for (y = y1, xx = x1; y <= y0; y++) {
                x = Math.trunc(xx);
                if (x >= 0 && x < img.getXSize() && y >= 0 && y < img.getYSize()) {
                    img.putPixelRgb(x, y, p);
                }
                xx += dxdy;
            }
        }
    }

    /**
    Draws a lines outline from a given polygon.
    */
    public static drawPolygon(img: Image, p: Polygon2D, color: RGBPixel): void {
        let va: Vertex2D;
        let vb: Vertex2D | null = null;
        let i: number;
        let j: number;

        for (i = 0; i < p.loops.length; i++) {
            const contour: _Polygon2DContour = p.loops[i]!;
            if (contour.vertices.length < 2) {
                continue;
            }
            for (j = 0; j < contour.vertices.length - 1; j++) {
                va = contour.vertices[j]!;
                vb = contour.vertices[j + 1]!;
                Rasterizer2D.drawLine(
                    img,
                    Math.trunc(va.x),
                    Math.trunc(va.y),
                    Math.trunc(vb.x),
                    Math.trunc(vb.y),
                    color,
                );
            }
            va = vb!;
            vb = contour.vertices[0]!;
            Rasterizer2D.drawLine(img, Math.trunc(va.x), Math.trunc(va.y), Math.trunc(vb.x), Math.trunc(vb.y), color);
        }
    }

    private static clamp(value: number, minValue: number, maxValue: number): number {
        if (value < minValue) {
            return minValue;
        }
        if (value > maxValue) {
            return maxValue;
        }
        return value;
    }

    private static addFillEdge(
        buckets: FillEdge[][],
        a: Vertex2D,
        b: Vertex2D,
        imageHeight: number,
        yRange: number[],
        sortOrder: number,
    ): void {
        let dx = b.x - a.x;
        let dy = b.y - a.y;

        if (Math.abs(dx) < VSDK.EPSILON && Math.abs(dy) < VSDK.EPSILON) {
            return;
        }
        if (Math.abs(dy) < VSDK.EPSILON) {
            return;
        }

        let top = a;
        let bottom = b;
        if (a.y > b.y) {
            top = b;
            bottom = a;
            dx = -dx;
            dy = -dy;
        }

        const inverseSlope = dx / dy;
        const yMin = Math.trunc(Math.ceil(top.y));
        const yMaxExclusive = Math.trunc(Math.ceil(bottom.y));

        if (yMin >= yMaxExclusive) {
            return;
        }

        const clippedYMin = Rasterizer2D.clamp(yMin, 0, imageHeight);
        const clippedYMaxExclusive = Rasterizer2D.clamp(yMaxExclusive, 0, imageHeight);

        if (clippedYMin >= clippedYMaxExclusive) {
            return;
        }

        const edge = new FillEdge();
        edge.yMin = clippedYMin;
        edge.yMaxExclusive = clippedYMaxExclusive;
        edge.inverseSlope = inverseSlope;
        edge.xAtCurrentY = top.x + (clippedYMin - top.y) * inverseSlope;
        edge.sortOrder = sortOrder;

        buckets[clippedYMin]!.push(edge);
        yRange[0] = Math.min(yRange[0]!, clippedYMin);
        yRange[1] = Math.max(yRange[1]!, clippedYMaxExclusive);
    }

    private static rasterizePolygonSpans(img: Image, polygon: Polygon2D, shader: SpanShader): void {
        const imageWidth = img.getXSize();
        const imageHeight = img.getYSize();

        if (imageWidth <= 0 || imageHeight <= 0) {
            return;
        }

        const buckets: FillEdge[][] = [];
        let y: number;
        for (y = 0; y < imageHeight; y++) {
            buckets.push([]);
        }

        const yRange: number[] = [imageHeight, 0];
        let sortOrder = 0;
        let i: number;
        let j: number;

        for (i = 0; i < polygon.loops.length; i++) {
            const contour: _Polygon2DContour = polygon.loops[i]!;
            const vertexCount = contour.vertices.length;
            if (vertexCount < 2) {
                continue;
            }

            for (j = 0; j < vertexCount; j++) {
                const a: Vertex2D = contour.vertices[j]!;
                const b: Vertex2D = contour.vertices[(j + 1) % vertexCount]!;
                Rasterizer2D.addFillEdge(buckets, a, b, imageHeight, yRange, sortOrder);
                sortOrder++;
            }
        }

        if (yRange[0]! >= yRange[1]!) {
            return;
        }

        const activeEdges: FillEdge[] = [];

        for (y = yRange[0]!; y < yRange[1]!; y++) {
            const bucket: FillEdge[] = buckets[y]!;
            if (bucket.length !== 0) {
                activeEdges.push(...bucket);
            }

            for (i = activeEdges.length - 1; i >= 0; i--) {
                if (y >= activeEdges[i]!.yMaxExclusive) {
                    activeEdges.splice(i, 1);
                }
            }

            if (activeEdges.length < 2) {
                for (i = 0; i < activeEdges.length; i++) {
                    activeEdges[i]!.xAtCurrentY += activeEdges[i]!.inverseSlope;
                }
                continue;
            }

            Collections.sort(activeEdges, Rasterizer2D.ACTIVE_EDGE_COMPARATOR);

            for (i = 0; i + 1 < activeEdges.length; i += 2) {
                let xLeft = activeEdges[i]!.xAtCurrentY;
                let xRight = activeEdges[i + 1]!.xAtCurrentY;

                if (xLeft > xRight) {
                    const tmp = xLeft;
                    xLeft = xRight;
                    xRight = tmp;
                }

                let xStart = Math.trunc(Math.ceil(xLeft));
                let xEndExclusive = Math.trunc(Math.ceil(xRight));

                if (xEndExclusive <= 0 || xStart >= imageWidth) {
                    continue;
                }

                xStart = Rasterizer2D.clamp(xStart, 0, imageWidth);
                xEndExclusive = Rasterizer2D.clamp(xEndExclusive, 0, imageWidth);

                if (xStart < xEndExclusive) {
                    shader.shade(img, polygon, y, xStart, xEndExclusive);
                }
            }

            for (i = 0; i < activeEdges.length; i++) {
                activeEdges[i]!.xAtCurrentY += activeEdges[i]!.inverseSlope;
            }
        }
    }

    /**
    Current polygon filling rasterizer (scan-line) algorithm is a NAIVE
    implementation of the general macro-algorithm outlined at [FOLE1992].3.6.

    This implementation is working but inefficient, due to the fact that
    neither scanline coherence nor edge coherence is taken into account. This
    means that for each scanline, all polygon edges are intersected
    analitically. This is the "brute-force technique" recommended to be
    avoided in [FOLE1992].3.6.3, but it is provided as reference to
    compare efficiency with other (clever) aproaches that makes use of
    mid-point algoritms for intersection finding, or "active edge tables"
    (AETs) for efficient edge-coherence based traversals.
    */
    public static fillPolygon(img: Image, p: Polygon2D, color: RGBPixel): void {
        Rasterizer2D.rasterizePolygonSpans(img, p, {
            shade(image: Image, polygon: Polygon2D, y: number, xStart: number, xEndExclusive: number): void {
                let x: number;
                for (x = xStart; x < xEndExclusive; x++) {
                    image.putPixelRgb(x, y, color);
                }
            },
        });
    }

    /**
    Given the current polygon `p` and the coordinate of a pixel (x, y),
    this method gives the interpolated color `outColor`.
    */
    public static fillSmoothPolygonCalculateColor(p: Polygon2D, x: number, y: number, outPixel: RGBPixel): void {
        let va: Vertex2D;
        let i: number;
        let j: number;
        let distance: number;
        let totaldistance = 0.0;

        let outR = 0.0;
        let outG = 0.0;
        let outB = 0.0;

        for (i = 0; i < p.loops.length; i++) {
            for (j = 0; j < p.loops[i]!.vertices.length; j++) {
                va = p.loops[i]!.vertices[j]!;
                distance = 1.0 / (1.0 + Math.sqrt((va.x - x) * (va.x - x) + (va.y - y) * (va.y - y)));
                totaldistance += distance;
                outR += va.color.r() * distance;
                outG += va.color.g() * distance;
                outB += va.color.b() * distance;
            }
        }

        const normalizedR = outR / totaldistance;
        const normalizedG = outG / totaldistance;
        const normalizedB = outB / totaldistance;
        const clippedR = Math.max(0.0, Math.min(1.0, normalizedR));
        const clippedG = Math.max(0.0, Math.min(1.0, normalizedG));
        const clippedB = Math.max(0.0, Math.min(1.0, normalizedB));
        const rr = Math.trunc(clippedR * 255.0);
        const gg = Math.trunc(clippedG * 255.0);
        const bb = Math.trunc(clippedB * 255.0);

        outPixel.r = VSDK.unsigned8BitInteger2signedByte(rr);
        outPixel.g = VSDK.unsigned8BitInteger2signedByte(gg);
        outPixel.b = VSDK.unsigned8BitInteger2signedByte(bb);
    }

    /**
    Current polygon filling rasterizer (scan-line) algorithm is a NAIVE
    implementation of the general macro-algorithm outlined at [FOLE1992].3.6.

    This implementation is working but inefficient, due to the fact that
    neither scanline coherence nor edge coherence is taken into account. This
    means that for each scanline, all polygon edges are intersected
    analitically. This is the "brute-force technique" recommended to be
    avoided in [FOLE1992].3.6.3, but it is provided as reference to
    compare efficiency with other (clever) aproaches that makes use of
    mid-point algoritms for intersection finding, or "active edge tables"
    (AETs) for efficient edge-coherence based traversals.
    */
    public static fillSmoothPolygon(img: Image, p: Polygon2D): void {
        Rasterizer2D.rasterizePolygonSpans(img, p, {
            shade(image: Image, polygon: Polygon2D, y: number, xStart: number, xEndExclusive: number): void {
                const color = new RGBPixel();
                let x: number;
                for (x = xStart; x < xEndExclusive; x++) {
                    Rasterizer2D.fillSmoothPolygonCalculateColor(polygon, x, y, color);
                    image.putPixelRgb(x, y, color);
                }
            },
        });
    }
}
